packages/snaps-controllers/src/insights/SnapInsightsController.ts

```diff
@@ -0,0 +1,378 @@
+        if (!state.insights[id]) {
+          state.insights[id] = {};
+        }
+        state.insights[id][snapId] = { snapId, loading: true };
+      });
+
+      this.#handleSnapRequest({
+        snapId,
+        handler: HandlerType.OnTransaction,
+        params: { transaction: txParams, chainId, transactionOrigin },
+      })
+        .then((response) =>
+          this.#handleSnapResponse({
+            id,
+            snapId,
+            response: response as Record<string, Json>,
+          }),
+        )
+        .catch((error) => this.#handleSnapResponse({ id, snapId, error }));
+    });
+  }
+
+  /**
+   * Handle the stateChange event emitted by the SignatureController.
+   * This function will remove existing insights from the state when applicable, as well as
+   * trigger insight fetching for newly added signatures.
+   *
+   * @param state - The SignatureController state blob.
+   */
+  #handleSignatureStateChange(state: SignatureControllerState) {
+    // If any IDs have disappeared since the last state update, the insight may be cleaned up.
+    for (const id of Object.keys(this.state.insights)) {
+      if (
+        !hasProperty(state.unapprovedTypedMessages, id) &&
+        !hasProperty(state.unapprovedPersonalMsgs, id)
+      ) {
+        this.#handleInsightCleanup(id);
+      }
+    }
+
+    if (
+      state.unapprovedPersonalMsgCount > 0 ||
+      state.unapprovedTypedMessagesCount > 0
+    ) {
+      const snaps = this.#getSnapsWithPermission(
+        SnapEndowments.SignatureInsight,
+      );
+
+      // This isn't very efficient, but SignatureController doesn't expose a better event for us to use yet.
+      for (const personalSignature of Object.values(
+        state.unapprovedPersonalMsgs,
+      )) {
+        this.#handleSignature(snaps, personalSignature);
+      }
+
+      for (const typedMessage of Object.values(state.unapprovedTypedMessages)) {
+        this.#handleSignature(snaps, typedMessage);
+      }
+    }
+  }
+
+  /**
+   * Handle a newly added unapproved signature.
+   * This function fetches insights from all available Snaps
+   * and populates the insights state blob with the responses.
+   *
+   * @param snaps - A list of Snaps to invoke.
+   * @param signature - The signature object.
+   */
+  #handleSignature(snaps: SnapWithPermission[], signature: StateSignature) {
+    const { id, msgParams } = signature;
+
+    if (this.#hasInsight(id)) {
+      return;
+    }
+
+    const { from, data, signatureMethod, origin } = msgParams;
+
+    /**
+     * Both eth_signTypedData_v3 and eth_signTypedData_v4 methods
+     * need to be parsed because their data is stringified. All other
+     * signature methods do not, so they are ignored.
+     */
+    const shouldParse =
+      signatureMethod === 'eth_signTypedData_v3' ||
+      signatureMethod === 'eth_signTypedData_v4';
+
+    const payload = {
+      from,
+      data: shouldParse ? JSON.parse(data as string) : data,
```

Comment:
Is `data` guaranteed to be a valid JSON string here?

Reply:
Seems to be assumed in the SignatureController as well yeah 🤔